### PR TITLE
Homework_query_1

### PR DESCRIPTION
**Pull request by raj-gupta09**

### CODE REVIEW

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. However, there are a few areas where it could be improved. Here are my suggestions:

1) **Covering all aspects of the homework prompt:** Your solution does not seem to handle the case where a player enters the league (`New`). The condition `(first_active_season - last_active_season) = 0` in your case statement does not accurately capture this scenario. Instead, you could use a condition like `l.player_name IS NULL AND c.player_name IS NOT NULL` to check if a player is new.

2) **Readability:** Your code is generally well-formatted and easy to read. However, it would be beneficial to include comments explaining your thought process and the purpose of each section of your code. This will make it easier for others (and future you) to understand what your code is doing.

3) **Robustness and error handling:** Your code does not seem to handle the case where a player stays out of the league (`Stayed Retired`). You could add a condition like `WHEN t.season IS NULL AND y.last_active_season < y.season` to handle this scenario.

4) **Modularity and reusability:** Your code is modular and each section is responsible for a specific task. However, the `final` CTE could be broken down into smaller parts to improve readability and reusability. For example, you could create separate CTEs for `last_season` and `current_season` before joining them in the `final` CTE.

5) **Variable names:** Your variable names are meaningful and descriptive, which makes your code easier to understand. However, the names `last_season` and `current_season_cte` could be more consistent. Consider renaming `current_season_cte` to `current_season` for consistency.

In conclusion, your solution is on the right track, but there are a few areas where it could be improved. Keep up the good work and continue to practice your SQL skills!

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly used the `GROUPING SETS` to perform aggregations of the `game_details` data. However, there are a few areas where your solution could be improved.

1) **Evaluate if the student's code covers all aspects of the homework prompt:**

Your code covers most aspects of the prompt, but there is a slight discrepancy in the way you've handled the `GROUPING` function. In the example solution, the `GROUPING` function is used with two arguments, while in your solution, it is used with a single argument. This might lead to incorrect results. 

2) **Assess if the student's solution logically follows the given homework prompt:**

Your solution logically follows the prompt. You've correctly used a CTE to prepare the data and then used `GROUPING SETS` to aggregate the data. However, the `GROUPING` function in your `CASE` statement should be used with two arguments to correctly identify the aggregation level.

3) **Assess the readability of their code:**

Your code is generally well-formatted and easy to read. However, you could improve the readability by adding comments to explain your thought process and the purpose of each part of your code. This would make it easier for others to understand your solution.

4) **Assess the robustness and error handling of the code:**

Your code seems to handle the given data well, but it's not clear how it would handle edge cases or invalid inputs. It would be beneficial to add some error handling or checks to ensure the data is as expected before performing the aggregations.

5) **Examine the modularity and reusability of the code:**

Your code is modular and the use of a CTE makes it easy to understand and reuse specific components. However, the `GROUPING` function in your `CASE` statement is not as flexible as it could be. Using two arguments, as in the example solution, would make this part of your code more reusable.

Here's a suggested modification to your code:

```sql
SELECT 
    CASE 
        WHEN GROUPING(player_name, team) = 0 THEN 'player_plus_team'
        WHEN GROUPING(player_name, season) = 0 THEN 'player_plus_season'
        WHEN GROUPING(team) = 0 THEN 'team'
        ELSE 'Overall'
    END AS Agg_Level,
    ...
```

Keep up the good work, and continue to focus on the details to ensure your code is robust and reusable.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly used `GROUPING SETS` to perform aggregations of the `game_details` data along the required dimensions. However, there are a few areas where your solution could be improved.

1) **Code Readability**: Your code is generally well-formatted and easy to read. You have used meaningful variable names and included comments explaining your thought process, which is good practice. However, the comments could be more detailed, especially for complex sections of the code. For example, you could explain why you are using `ROW_NUMBER()` and how it affects the final result.

2) **Robustness and Error Handling**: Your code does not seem to handle potential errors or edge cases. For example, what happens if the `game_id`, `team_id`, or `player_id` is null? You should consider adding error handling or checks for invalid inputs to make your code more robust.

3) **Modularity and Reusability**: Your code is structured in a way that makes it easy to understand and reuse specific components in other contexts. The use of a CTE (Common Table Expression) at the beginning of your query is a good practice as it improves readability and allows for code reuse.

4) **Missing or Incomplete Parts**: Your solution does not exactly match the example solution, but it still meets the requirements of the homework prompt. However, there is a slight discrepancy in the way you calculate `total_win`. In the example solution, the `won` column is calculated in the `deduped` CTE and then summed up in the final `SELECT` statement. In your solution, you calculate `total_win` directly in the `games_grouping` CTE. While this does not affect the final result, it could make your code harder to understand and maintain in the long run.

To improve your solution, I would suggest adding more detailed comments, implementing error handling, and considering the maintainability of your code when calculating aggregated columns.

### Feedback for `query_3.sql`

Overall, your solution is well-structured and logically follows the homework prompt. You have successfully written a query to find the player who scored the most points playing for a single team. However, there are a few areas where your solution could be improved.

1) **Readability and Comments:** Your code is generally well-commented, which makes it easier to understand your thought process. However, the comments could be more concise and to the point. For example, the comment `-- Selecting player name, team, and total points from the specified table` could be simplified to `-- Select player, team, and points`.

2) **Variable Names:** The variable names you've chosen are mostly clear and descriptive. However, the `Agg_Level` column name might be more readable if it were written in lowercase, as `agg_level`, to follow SQL naming conventions. Also, it would be better if it matches the example solution's `aggregation_level` for consistency.

3) **Robustness and Error Handling:** Your code does not include any explicit error handling or checks for edge cases. While not always necessary in SQL, it can be helpful to include checks for unexpected or invalid inputs. For example, you could add a check to ensure that `total_player_points` is a positive number.

4) **Modularity and Reusability:** Your use of a CTE (Common Table Expression) with the `WITH` clause is a good practice for creating modular, reusable code. However, the CTE might be unnecessary in this case, as the query could be written more simply without it. Consider whether the added complexity of the CTE is necessary for this particular problem.

Here's a revised version of your code incorporating these suggestions:

```sql
-- Selecting player name, team, and total points
SELECT
    player_name,
    team,
    total_player_points
FROM
    raj.nba_games_grouping
WHERE
    -- Filtering for rows representing a single player and a single team 
    aggregation_level = 'player_name__team_name'
    AND total_player_points IS NOT NULL
    AND total_player_points > 0
ORDER BY
    -- Sorting by points in descending order
    total_player_points DESC
LIMIT 1 
```

Keep up the good work, and continue to focus on writing clear, concise, and robust code!

### Feedback for `query_4.sql`

Overall, your code is well-structured, readable, and logically follows the homework prompt. You've done a good job of using a CTE (Common Table Expression) to filter the data before selecting the top-scoring player. This makes your code more modular and easier to understand. 

However, there are a few areas where your solution could be improved:

1) **Homework Prompt Coverage:** The homework prompt asks for the player who scored the most points in one season, but your code does not include the 'season' in the output. You should modify your code to include the 'season' in both the CTE and the final SELECT statement. 

2) **Readability:** Your code is generally well-formatted and easy to read. However, you could improve the readability by adding more comments to explain your thought process, especially in the CTE where you are filtering the data.

3) **Robustness and Error Handling:** Your code does a good job of handling potential errors by filtering out rows where 'total_player_points' is NULL. However, you should also consider handling potential errors related to the 'player_name' and 'season' fields. For example, you could add a condition to filter out rows where 'player_name' or 'season' is NULL.

4) **Modularity and Reusability:** Your use of a CTE makes your code more modular and easier to understand. However, you could improve the reusability of your code by parameterizing the 'Agg_Level' value. This would allow you to easily modify the code to select the top-scoring player for different aggregation levels.

Here's how you could modify your code to address these points:

```sql
WITH player_season_points AS (
    -- Selecting player name, season, team, and total points from the specified table
    SELECT
        player_name,
        season,
        team,
        total_player_points
    FROM
        raj.nba_games_grouping
    WHERE
        -- Filtering for rows representing a combination of a single player and a season and where points are available
        Agg_Level = 'player_plus_season'
        AND total_player_points IS NOT NULL
        AND player_name IS NOT NULL
        AND season IS NOT NULL
)
-- Selecting the top-scoring record from the filtered results
SELECT
    player_name,
    season,
    team,
    total_player_points
FROM
    player_season_points
ORDER BY
    total_player_points DESC      -- Sorting the rows in descending order based on the points to bring the highest scorer to the top
LIMIT 1                           -- Return only the top scoring record
```

Keep up the good work! With a few minor adjustments, your code will be even better.

### Feedback for `query_5.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have successfully written a query to find the team with the most wins, which is the main requirement of the prompt. However, there are a few areas where your solution could be improved.

1) **Readability and Comments:** Your code is generally readable, and your use of comments is commendable. However, the comments could be more concise and still convey the same information. For example, the comment "Selecting team name and total wins from the specified table" could be shortened to "Select team name and total wins". Similarly, "Filtering for rows representing a single team and where wins are available" could be shortened to "Filter rows for single team with available wins". 

2) **Robustness and Error Handling:** Your code does not handle potential errors or edge cases. For instance, what happens if there are multiple teams with the same highest number of wins? Your query will only return one of them. To handle this, you could remove the `LIMIT 1` clause and instead use a `HAVING` clause to only include teams whose total wins are equal to the maximum total wins.

3) **Modularity and Reusability:** Your code is modular and could be reused in other contexts with minor modifications. However, the `WITH` clause is not necessary in this case. You could simplify your query by removing the `WITH` clause and including the `WHERE` clause conditions directly in the main `SELECT` statement.

Here's a revised version of your query incorporating these suggestions:

```sql
-- Select the top team(s) based on total wins
SELECT
    team,
    total_win
FROM
    raj.nba_games_grouping
WHERE
    -- Filter rows for single team with available wins
    Agg_Level = 'team'
    AND total_win IS NOT NULL
HAVING
    total_win = MAX(total_win)
ORDER BY
    total_win DESC
```

Keep up the good work, and continue practicing to improve your SQL skills!

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_6.sql`

Overall, your solution is well-structured and logically follows the assignment prompt. However, there are a few areas where it could be improved. Here are my suggestions:

1) **Assignment Coverage:** The prompt asks for the most games a single team has won in a given 90-game stretch, but your solution calculates the most games won in a 90-day stretch. This is a significant difference, as a team could play multiple games in a single day. You should adjust your window function to count the number of games, not the number of days. 

2) **Readability:** Your code is generally well-structured and easy to read. You've used meaningful variable names and included comments to explain your thought process, which is great. However, the `home_team_id` and `visitor_team_id` variables in the `team_games` CTE could be confusing. Consider renaming these to `team_id` and `opponent_id` to better reflect their purpose in this context.

3) **Robustness and Error Handling:** Your code does not seem to handle the case where a game ends in a tie. While ties are rare in basketball, they can occur, and your code should account for this possibility. You could add a condition to handle this case in your `team_games` CTE.

4) **Modularity and Reusability:** Your code is modular and each CTE serves a specific purpose, which is good. However, the `distinct_games` CTE seems unnecessary, as the `nba_games` table should not have duplicate entries. Removing this CTE would simplify your code and make it more efficient.

Here's a revised version of your code with these suggestions:

```sql
WITH team_games AS (
    -- Combine home and away games for each team
    SELECT
        game_date_est AS game_date,
        home_team_id AS team_id,
        CASE WHEN home_team_wins = 1 THEN 1 ELSE 0 END AS is_won
    FROM
        bootcamp.nba_games
    UNION ALL
    SELECT
        game_date_est AS game_date,
        visitor_team_id AS team_id,
        CASE WHEN home_team_wins = 0 THEN 1 ELSE 0 END AS is_won
    FROM
        bootcamp.nba_games
),

wins_90_games AS (
    -- Compute 90-game sliding window win totals
    SELECT
        team_id,
        game_date AS window_end,
        SUM(is_won) OVER (
            PARTITION BY team_id
            ORDER BY game_date
            ROWS BETWEEN 89 PRECEDING AND CURRENT ROW
        ) AS number_of_wins
    FROM
        team_games
)

-- Select team with max wins in a 90-game window
SELECT
    max_by(team_id, number_of_wins) AS team_id,
    max_by(window_end, number_of_wins) AS window_end,
    MAX(number_of_wins) AS most_90_games_wins
FROM
    wins_90_games
```

Keep up the good work! With a few adjustments, your solution will be even better.

### Feedback for `query_7.sql`

The student's solution is well-structured and logically follows the given homework prompt. The student has used window functions to answer the question: "How many games in a row did LeBron James score over 10 points a game?" The code is divided into several CTEs (Common Table Expressions), which makes it easier to understand the logic behind each step.

The student's code is readable and well-commented. The variable names are meaningful and descriptive, which makes it easier to understand what each part of the code is doing. The use of comments to explain the purpose of each CTE is also commendable.

The student's solution is robust and handles edge cases well. The use of the `LAG` function to calculate the lagged value of the 'ten_plus' flag is a good approach to identify streaks of consecutive games where LeBron James scored more than 10 points. The use of the `SUM` function in the `streak_identification` CTE to increment the streak_id whenever the 'ten_plus' flag changes is a clever way to identify individual streaks.

However, there is a potential issue with the student's solution. The `MAX_BY` function in the final `SELECT` statement might not return the correct player_id and player_name if there are multiple players with the same maximum streak length. Since the homework prompt only asks about LeBron James, this might not be a problem in this specific case, but it's something to keep in mind for more general queries.

The student's code is modular and reusable. The CTEs can be easily modified or reused in other contexts. However, the final `SELECT` statement is specific to this particular question and might not be as easily reusable.

Overall, the student's solution is well-done and demonstrates a good understanding of SQL window functions. The student could improve their solution by addressing the potential issue with the `MAX_BY` function in the final `SELECT` statement. One possible solution would be to add a `WHERE` clause to the `streak_lengths` CTE to only include rows where player_name = 'LeBron James'. This would ensure that the final `SELECT` statement only returns the maximum streak length for LeBron James.
